File: src/main/java/com/yupi/template/job/SyncModelJob.java
package com.yupi.template.job;

import cn.hutool.json.JSONUtil;
import com.yupi.template.constant.ConversationConstant;
import com.yupi.template.mapper.ModelMapper;
import com.yupi.template.model.dto.openrouter.OpenRouterModelResponse;
import com.yupi.template.model.entity.Model;
import com.yupi.template.service.ModelService;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * 同步OpenRouter模型列表定时任务
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@Component
@Slf4j
public class SyncModelJob {

    @Resource
    private ModelMapper modelMapper;

    @Resource
    private ModelService modelService;

    @Value("${spring.ai.openai.api-key}")
    private String openRouterApiKey;

    private static final String OPENROUTER_MODELS_URL = "https://openrouter.ai/api/v1/models";


    /**
     * 每天凌晨2点执行一次
     * 可以通过配置文件覆盖：job.sync-model.cron
     */
    @Scheduled(cron = "0 0 2 * * ?")
    public void syncModels() {
        log.info("开始同步OpenRouter模型列表");

        try {
            // 调用OpenRouter API获取模型列表
            List<OpenRouterModelResponse.OpenRouterModel> openRouterModels = fetchModelsFromOpenRouter();

            if (openRouterModels == null || openRouterModels.isEmpty()) {
                log.warn("未获取到任何模型数据");
                return;
            }

            log.info("从OpenRouter获取到{}个模型", openRouterModels.size());

            // 转换并保存到数据库
            int successCount = 0;
            int failCount = 0;

            for (OpenRouterModelResponse.OpenRouterModel openRouterModel : openRouterModels) {
                try {
                    Model model = convertToModel(openRouterModel);

                    // 使用replace方式：存在则更新，不存在则插入
                    Model existingModel = modelMapper.selectOneById(model.getId());
                    if (existingModel != null) {
                        model.setCreateTime(existingModel.getCreateTime());
                        modelMapper.update(model);
                    } else {
                        modelMapper.insert(model);
                    }
                    modelService.evictModelPricingCache(model.getId());

                    successCount++;
                } catch (Exception e) {
                    log.error("保存模型失败: {}", openRouterModel.getId(), e);
                    failCount++;
                }
            }

            log.info("模型同步完成: 成功{}个, 失败{}个", successCount, failCount);
        } catch (Exception e) {
            log.error("同步模型列表失败", e);
        }
    }

    /**
     * 从OpenRouter API获取模型列表
     */
    private List<OpenRouterModelResponse.OpenRouterModel> fetchModelsFromOpenRouter() {
        try {
            RestTemplate restTemplate = new RestTemplate();

            // 设置请求头
            HttpHeaders headers = new HttpHeaders();
            headers.set("Authorization", "Bearer " + openRouterApiKey);
            HttpEntity<String> entity = new HttpEntity<>(headers);

            // 调用API
            ResponseEntity<OpenRouterModelResponse> response = restTemplate.exchange(
                    OPENROUTER_MODELS_URL,
                    HttpMethod.GET,
                    entity,
                    OpenRouterModelResponse.class
            );

            if (response.getBody() != null && response.getBody().getData() != null) {
                return response.getBody().getData();
            }
        } catch (Exception e) {
            log.error("调用OpenRouter API失败", e);
        }

        return null;
    }

    /**
     * 转换OpenRouter模型为数据库实体
     */
    private Model convertToModel(OpenRouterModelResponse.OpenRouterModel openRouterModel) {
        // 提取提供商
        String provider = extractProvider(openRouterModel.getId());

        // 转换价格（从每token转为每百万tokens）
        BigDecimal inputPrice = convertPrice(openRouterModel.getPricing().getPrompt());
        BigDecimal outputPrice = convertPrice(openRouterModel.getPricing().getCompletion());


        // 生成标签
        String[] tags = generateTags(openRouterModel);
        String tagsJson = JSONUtil.toJsonStr(tags);

        // 保存原始数据
        String rawData = JSONUtil.toJsonStr(openRouterModel);

        LocalDateTime now = LocalDateTime.now();
        return Model.builder()
                .id(openRouterModel.getId())
                .name(openRouterModel.getName())
                .description(openRouterModel.getDescription())
                .provider(provider)
                .contextLength(openRouterModel.getContextLength())
                .inputPrice(inputPrice)
                .outputPrice(outputPrice)
                .recommended(0)
                .tags(tagsJson)
                .rawData(rawData)
                .createTime(now)
                .updateTime(now)
                .isDelete(0)
                .build();
    }

    /**
     * 从模型ID提取提供商名称
     */
    private String extractProvider(String modelId) {
        if (modelId == null || !modelId.contains("/")) {
            return "Unknown";
        }
        String prefix = modelId.split("/")[0];
        return switch (prefix.toLowerCase()) {
            // 国产厂商
            case "qwen", "alibaba" -> "Alibaba";
            case "deepseek" -> "DeepSeek";
            case "z-ai", "zhipu" -> "Zhipu AI";
            case "moonshotai" -> "Moonshotai";
            case "baidu" -> "Baidu";
            case "tencent" -> "Tencent";
            case "bytedance", "bytedance-seed" -> "Bytedance";
            case "meituan" -> "Meituan";
            // 国外厂商
            case "openai" -> "OpenAI";
            case "anthropic" -> "Anthropic";
            case "google" -> "Google";
            case "meta-llama", "meta" -> "Meta";
            case "mistralai" -> "Mistral AI";
            default -> capitalize(prefix);
        };
    }

    /**
     * 首字母大写
     */
    private String capitalize(String str) {
        if (str == null || str.isEmpty()) {
            return str;
        }
        return str.substring(0, 1).toUpperCase() + str.substring(1);
    }

    /**
     * 转换价格：从每token价格转换为每百万tokens价格
     */
    private BigDecimal convertPrice(String priceStr) {
        try {
            if (priceStr == null || priceStr.isEmpty()) {
                return BigDecimal.ZERO;
            }
            // OpenRouter返回的价格是每token的美元价格（科学计数法字符串）
            BigDecimal pricePerToken = new BigDecimal(priceStr);
            return pricePerToken.multiply(new BigDecimal(ConversationConstant.TOKENS_PER_MILLION));
        } catch (Exception e) {
            log.warn("价格转换失败: {}", priceStr, e);
            return BigDecimal.ZERO;
        }
    }

    /**
     * 生成模型标签
     */
    private String[] generateTags(OpenRouterModelResponse.OpenRouterModel model) {
        List<String> tags = new ArrayList<>();

        // 根据模态添加标签
        if (model.getArchitecture() != null) {
            if (model.getArchitecture().getInputModalities() != null) {
                if (model.getArchitecture().getInputModalities().contains("image")) {
                    tags.add("多模态");
                }
            }
            if (model.getArchitecture().getOutputModalities() != null) {
                if (model.getArchitecture().getOutputModalities().contains("image")) {
                    tags.add("图像生成");
                }
            }
        }

        // 根据模型名称添加标签
        String name = model.getName().toLowerCase();
        String id = model.getId().toLowerCase();

        if (name.contains("code") || id.contains("code")) {
            tags.add("代码");
        }
        if (name.contains("mini") || name.contains("flash") || name.contains("haiku")) {
            tags.add("快速");
        }
        if (name.contains("vision") || name.contains("multimodal")) {
            tags.add("多模态");
        }
        if (id.contains("qwen") || id.contains("glm") || id.contains("ernie")) {
            tags.add("中文");
        }

        // 默认添加文本生成标签
        if (tags.isEmpty()) {
            tags.add("文本生成");
        } else if (!tags.contains("文本生成")) {
            tags.add("文本生成");
        }

        return tags.toArray(new String[0]);
    }
}

File: go-backend/internal/job/sync_model_job.go
// Package job 同步OpenRouter模型列表定时任务
// @author <a href="https://codefather.cn">编程导航学习圈</a>
package job

import (
	"ai-test-go/internal/config"
	"ai-test-go/internal/model"
	"ai-test-go/internal/model/dto"
	"ai-test-go/internal/repository"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/robfig/cron/v3"
)

const (
	OPENROUTER_MODELS_URL = "https://openrouter.ai/api/v1/models"
	TOKENS_PER_MILLION    = 1000000
)

type ModelPricingCacheEvictor interface {
	EvictModelPricingCache(modelName string)
}

type SyncModelJob struct {
	modelRepo   *repository.ModelRepository
	cacheEvict  ModelPricingCacheEvictor
	cron        *cron.Cron
}

func NewSyncModelJob(modelRepo *repository.ModelRepository, cacheEvict ModelPricingCacheEvictor) *SyncModelJob {
	return &SyncModelJob{
		modelRepo:  modelRepo,
		cacheEvict: cacheEvict,
		cron:       cron.New(),
	}
}

func (j *SyncModelJob) Start() {
	j.cron.AddFunc("0 2 * * *", func() {
		j.SyncModels()
	})
	j.cron.Start()
	log.Println("模型同步定时任务已启动，每天凌晨2点执行")
}

func (j *SyncModelJob) Stop() {
	j.cron.Stop()
}

func (j *SyncModelJob) SyncModels() {
	log.Println("开始同步OpenRouter模型列表")

	openRouterModels, err := j.fetchModelsFromOpenRouter()
	if err != nil {
		log.Printf("获取OpenRouter模型列表失败: %v", err)
		return
	}

	if len(openRouterModels) == 0 {
		log.Println("未获取到任何模型数据")
		return
	}

	log.Printf("从OpenRouter获取到%d个模型", len(openRouterModels))

	successCount := 0
	failCount := 0

	for _, openRouterModel := range openRouterModels {
		m := j.convertToModel(openRouterModel)
		if err := j.modelRepo.SaveOrUpdate(m); err != nil {
			log.Printf("保存模型失败: %s, 错误: %v", openRouterModel.ID, err)
			failCount++
		} else {
			if j.cacheEvict != nil {
				j.cacheEvict.EvictModelPricingCache(openRouterModel.ID)
			}
			successCount++
		}
	}

	log.Printf("模型同步完成: 成功%d个, 失败%d个", successCount, failCount)
}

func (j *SyncModelJob) fetchModelsFromOpenRouter() ([]dto.OpenRouterModel, error) {
	client := &http.Client{Timeout: 30 * time.Second}

	req, err := http.NewRequest("GET", OPENROUTER_MODELS_URL, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", "Bearer "+config.AppConfig.OpenRouter.APIKey)

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("API请求失败: status=%d, body=%s", resp.StatusCode, string(body))
	}

	var response dto.OpenRouterModelResponse
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return nil, err
	}

	return response.Data, nil
}

func (j *SyncModelJob) convertToModel(openRouterModel dto.OpenRouterModel) *model.Model {
	provider := j.extractProvider(openRouterModel.ID)
	isChina := 0
	if j.isChinaModel(openRouterModel.ID) {
		isChina = 1
	}

	supportsMultimodal := 0
	if openRouterModel.Architecture.InputModalities != nil {
		for _, m := range openRouterModel.Architecture.InputModalities {
			if m == "image" {
				supportsMultimodal = 1
				break
			}
		}
	}

	supportsImageGen := 0
	if openRouterModel.Architecture.OutputModalities != nil {
		for _, m := range openRouterModel.Architecture.OutputModalities {
			if m == "image" {
				supportsImageGen = 1
				break
			}
		}
	}

	supportsToolCalling := 0
	if openRouterModel.SupportedParameters != nil {
		for _, param := range openRouterModel.SupportedParameters {
			if param == "tools" || param == "tool_choice" || param == "functions" || param == "function_call" {
				supportsToolCalling = 1
				break
			}
		}
	}

	inputPrice := j.convertPrice(openRouterModel.Pricing.Prompt)
	outputPrice := j.convertPrice(openRouterModel.Pricing.Completion)

	tags := j.generateTags(openRouterModel)
	tagsJSON, _ := json.Marshal(tags)

	rawData, _ := json.Marshal(openRouterModel)

	now := time.Now()
	return &model.Model{
		ID:                  openRouterModel.ID,
		Name:                openRouterModel.Name,
		Description:         openRouterModel.Description,
		Provider:            provider,
		ContextLength:       openRouterModel.ContextLength,
		InputPrice:          inputPrice,
		OutputPrice:         outputPrice,
		Recommended:         0,
		IsChina:             isChina,
		SupportsMultimodal:  supportsMultimodal,
		SupportsImageGen:    supportsImageGen,
		SupportsToolCalling: supportsToolCalling,
		Tags:                string(tagsJSON),
		RawData:             string(rawData),
		TotalTokens:         0,
		TotalCost:           0,
		CreateTime:          now,
		UpdateTime:          now,
		IsDelete:            0,
	}
}

func (j *SyncModelJob) isChinaModel(modelID string) bool {
	if !strings.Contains(modelID, "/") {
		return false
	}
	prefix := strings.ToLower(strings.Split(modelID, "/")[0])
	chinaProviders := []string{"qwen", "alibaba", "deepseek", "z-ai", "zhipu", "moonshotai", "baidu", "tencent", "bytedance", "bytedance-seed", "meituan"}
	for _, p := range chinaProviders {
		if prefix == p {
			return true
		}
	}
	return false
}

func (j *SyncModelJob) extractProvider(modelID string) string {
	if !strings.Contains(modelID, "/") {
		return "Unknown"
	}
	prefix := strings.ToLower(strings.Split(modelID, "/")[0])

	providerMap := map[string]string{
		"qwen":           "Alibaba",
		"alibaba":        "Alibaba",
		"deepseek":       "DeepSeek",
		"z-ai":           "Zhipu AI",
		"zhipu":          "Zhipu AI",
		"moonshotai":     "Moonshotai",
		"baidu":          "Baidu",
		"tencent":        "Tencent",
		"bytedance":      "Bytedance",
		"bytedance-seed": "Bytedance",
		"meituan":        "Meituan",
		"openai":         "OpenAI",
		"anthropic":      "Anthropic",
		"google":         "Google",
		"meta-llama":     "Meta",
		"meta":           "Meta",
		"mistralai":      "Mistral AI",
	}

	if provider, ok := providerMap[prefix]; ok {
		return provider
	}

	return capitalize(prefix)
}

func capitalize(str string) string {
	if len(str) == 0 {
		return str
	}
	return strings.ToUpper(str[:1]) + str[1:]
}

func (j *SyncModelJob) convertPrice(priceStr string) float64 {
	if priceStr == "" {
		return 0
	}
	price, err := strconv.ParseFloat(priceStr, 64)
	if err != nil {
		log.Printf("价格转换失败: %s", priceStr)
		return 0
	}
	return price * TOKENS_PER_MILLION
}

func (j *SyncModelJob) generateTags(m dto.OpenRouterModel) []string {
	tags := []string{}

	if m.Architecture.InputModalities != nil {
		for _, mod := range m.Architecture.InputModalities {
			if mod == "image" {
				tags = append(tags, "多模态")
				break
			}
		}
	}

	if m.Architecture.OutputModalities != nil {
		for _, mod := range m.Architecture.OutputModalities {
			if mod == "image" {
				tags = append(tags, "图像生成")
				break
			}
		}
	}

	nameLower := strings.ToLower(m.Name)
	idLower := strings.ToLower(m.ID)

	if strings.Contains(nameLower, "code") || strings.Contains(idLower, "code") {
		tags = append(tags, "代码")
	}
	if strings.Contains(nameLower, "mini") || strings.Contains(nameLower, "flash") || strings.Contains(nameLower, "haiku") {
		tags = append(tags, "快速")
	}
	if strings.Contains(nameLower, "vision") || strings.Contains(nameLower, "multimodal") {
		tags = append(tags, "多模态")
	}
	if strings.Contains(idLower, "qwen") || strings.Contains(idLower, "glm") || strings.Contains(idLower, "ernie") {
		tags = append(tags, "中文")
	}

	hasTextGen := false
	for _, tag := range tags {
		if tag == "文本生成" {
			hasTextGen = true
			break
		}
	}
	if !hasTextGen {
		tags = append(tags, "文本生成")
	}

	return tags
}
